### PR TITLE
Add `space` to JSON.stringify() for Node 0.10 issue

### LINKED ISSUE #3: slow serialization for some cases (happens in Node.js 10)
Hi @ericf as we discussed, there might be some case to cause a `GC pause`, 

```
I'm guessing it's some V8 de-opt we're hitting because the replacer forces V8 to go from native-land to user-land and back. Whereas with JSON.stringify().replace() it can stay in native-land the whole time.
So it's probably causing a serialization/de-serialization of the massive string causing a GC pause.
```

Now we have data `~500KB` to be serialized, and found that some case(typically when the data is complicated and big enough) [str.replace] (https://github.com/yahoo/serialize-javascript/blob/master/index.js#L56-L58) takes a very long time after [JSON.stringify with replacer](https://github.com/yahoo/serialize-javascript/blob/master/index.js#L35-L45)

To reproduce it, here's the simple script with the data we have,  
We will get rid of it with a `gc()` or run without `replacer`, it acts like what you mentioned. 
https://drive.google.com/file/d/0B0ZB5_LvayjrRjV0U2hrMGVjMUE/view?usp=sharing
please run `node --trace-gc --replacer --issue ./slowserialization.js` to reproduce it and see gc solve it with `node --trace-gc --expose-gc --replacer --issue --gc ./slowserialization.js`
 
will see when slow serialization happens, there's a lot of sweep and cause time spend. 
```
[41363]     3133 ms: Mark-sweep 331.0 (368.4) -> 327.2 (366.4) MB, 123 ms [last resort gc].
[41363]     3249 ms: Mark-sweep 327.2 (366.4) -> 327.2 (364.4) MB, 116 ms [last resort gc].
[41363]     3361 ms: Mark-sweep 327.2 (364.4) -> 327.2 (364.4) MB, 112 ms [last resort gc].
[41363]     3475 ms: Mark-sweep 327.2 (364.4) -> 327.2 (364.4) MB, 113 ms [last resort gc].
[41363]     3578 ms: Mark-sweep 327.2 (364.4) -> 327.2 (364.4) MB, 103 ms [last resort gc].
[41363]     3676 ms: Mark-sweep 327.2 (364.4) -> 327.2 (364.4) MB, 97 ms [last resort gc].
[41363]     3776 ms: Mark-sweep 327.2 (364.4) -> 327.2 (364.4) MB, 100 ms [last resort gc].
```
The slow serialization only happens with Node.js `10`, `12` won't have that issue.

I don't think that makes sense to fix this only for Node.js 10, but let's discuss here, like can we try something else for handling the `function/regexp` or make that  optional (for our main customer's case, we don't have function and regexp)?

@redonkulus @lingyan @mridgway @Vijar

### PR DESCRIPTION
**Pull request by ericf**
Fixes #3

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ See the accompanying LICENSE file for terms.
 
 'use strict';
 
-var util = require('util');
+var isRegExp = require('util').isRegExp;
```
Comment (ericf) [inline index.js]: Optimize away the property lookup in the replacer.

File: index.js [hunk 2]
```diff
@@ -37,12 +44,12 @@ function serialize(obj) {
             return '@__FUNCTION_' + (functions.push(value) - 1) + '__@';
         }
 
-        if (util.isRegExp(value)) {
+        if (typeof value === 'object' && isRegExp(value)) {
```
Comment (ericf) [inline index.js]: Type-check before function call.

File: index.js [hunk 3]
```diff
@@ -57,7 +64,7 @@ function serialize(obj) {
         return UNICODE_CHARS[unsafeChar];
     });
 
-    if (!(functions.length || regexps.length)) {
+    if (functions.length === 0 && regexps.length === 0) {
```
Comment (ericf) [inline index.js]: For clarity.

File: test/unit/serialize.js [hunk 4]
```diff
@@ -46,7 +46,7 @@ describe('serialize( obj )', function () {
         });
 
         it('should serialize JSON to a JSON string', function () {
-            expect(serialize(data)).to.equal(JSON.stringify(data));
+            expect(serialize(data)).to.equal(JSON.stringify(data, null, 2));
```
Comment (ericf) [inline test/unit/serialize.js]: Required now that we have spaces in the output.

**Comment by caridy** [comment]
LGTM

**Comment by Vijar** [comment]
:+1:

**Comment by kaesonho** [comment]
:+1:  thanks @ericf

**Comment by ericf** [comment]
This PR has been reverted.

**Pull request merged**